File: python/src/bursar/credits/postgres/repositories/pricing.py
from __future__ import annotations

import json

from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories.schemas import ActivePricingRow, BursarConfigHistoryItemRow


class PricingRepository:
    def __init__(self, callproc: DbQuery) -> None:
        self._callproc = callproc

    @staticmethod
    def _parse_revision(row: dict[str, object]) -> ActivePricingRow:
        parsed = dict(row)
        if parsed.get("id") is not None:
            parsed["id"] = str(parsed["id"])
        parsed.update(
            {
                "config": parsed.get("source_document"),
                "version": parsed.get("revision_no"),
                "active": parsed.get("status") == "active",
            }
        )
        return ActivePricingRow.model_validate(parsed)

    def get_active_pricing(self) -> ActivePricingRow | None:
        rows = self._callproc("active_catalog_revision", []) or []
        if not rows or not isinstance(rows[0], dict):
            return None
        row = dict(rows[0])
        if not row or all(value is None for value in row.values()):
            return None
        return self._parse_revision(row)

    def set_active_pricing(self, config: str, label: str | None) -> ActivePricingRow | None:
        return self._publish_pricing(config, label, activate=True)

    def publish_pricing(self, config: str, label: str | None) -> ActivePricingRow | None:
        return self._publish_pricing(config, label, activate=False)

    def _publish_pricing(self, config: str, label: str | None, *, activate: bool) -> ActivePricingRow | None:
        rows = self._callproc("publish_and_activate_catalog", [1, json.loads(config), label, activate]) or []
        if not rows or not isinstance(rows[0], dict) or rows[0].get("revision_no") is None:
            return None
        return self.get_bursar_config(int(rows[0]["revision_no"]))

    def get_pricing_history(self) -> list[BursarConfigHistoryItemRow]:
        rows = self._callproc("list_catalog_revisions", [500]) or []
        return [
            BursarConfigHistoryItemRow.model_validate(self._parse_revision(dict(row)).model_dump())
            for row in rows
            if isinstance(row, dict)
        ]

    def get_bursar_config(self, version: int) -> ActivePricingRow | None:
        rows = self._callproc("catalog_revision_by_number", [version]) or []
        if not rows or not isinstance(rows[0], dict):
            return None
        return self._parse_revision(dict(rows[0]))

    def activate_pricing(self, version: int) -> ActivePricingRow | None:
        rows = self._callproc("activate_catalog_revision", [version]) or []
        if not rows or not isinstance(rows[0], dict):
            return None
        return self._parse_revision(dict(rows[0]))

File: javascript/src/credits/postgres/repositories/pricing.ts
import { z } from "zod";
import type { CallProc } from "../../../shared/postgres-types.js";
import { pgBoolean, safeParse } from "../../../shared/postgres-validation.js";

const ActivePricingRowSchema = z
  .object({
    id: z.string().optional(),
    config: z.record(z.string(), z.unknown()).optional(),
    source_document: z.record(z.string(), z.unknown()).optional(),
    version: z.coerce.number().optional(),
    revision_no: z.coerce.number().optional(),
    label: z.string().nullable().optional(),
    active: pgBoolean.nullable().optional(),
    status: z.string().nullable().optional(),
    created_at: z.union([z.string(), z.date().transform((v) => v.toISOString())]).optional(),
  })
  .passthrough();
export type ActivePricingRow = z.infer<typeof ActivePricingRowSchema>;
export type PricingHistoryRow = ActivePricingRow;

export class PricingRepository {
  constructor(private readonly callproc: CallProc) {}

  private parseRevision(row: Record<string, unknown>, context: string): ActivePricingRow {
    return safeParse(
      ActivePricingRowSchema,
      {
        ...row,
        config: row.source_document,
        version: row.revision_no,
        active: row.status === "active",
      },
      context,
    );
  }

  async getActivePricing(): Promise<ActivePricingRow | null> {
    const rows = await this.callproc("active_catalog_revision", []);
    if (!rows?.length) return null;
    const row = rows[0] as Record<string, unknown>;
    return this.parseRevision(row, "PricingRepository.getActivePricing");
  }

  async setActivePricing(config: string, label: string | null): Promise<ActivePricingRow> {
    return this.publishPricing(config, label, true);
  }

  async publishPricing(
    config: string,
    label: string | null,
    activate = false,
  ): Promise<ActivePricingRow> {
    const published = (
      await this.callproc("publish_and_activate_catalog", [1, JSON.parse(config), label, activate])
    )[0] as Record<string, unknown> | undefined;
    if (published?.revision_no == null) {
      return safeParse(ActivePricingRowSchema, {}, "PricingRepository.publishPricing");
    }
    const revision = await this.getBursarConfig(Number(published.revision_no));
    return (
      revision ??
      safeParse(
        ActivePricingRowSchema,
        {
          id: published.revision_id,
          version: published.revision_no,
          status: published.status,
          active: published.status === "active",
        },
        "PricingRepository.publishPricing",
      )
    );
  }

  async getPricingHistory(): Promise<PricingHistoryRow[]> {
    const rows = await this.callproc("list_catalog_revisions", [500]);
    return (rows ?? []).map((row) =>
      this.parseRevision(row as Record<string, unknown>, "PricingRepository.getPricingHistory"),
    );
  }

  async getBursarConfig(version: number): Promise<ActivePricingRow | null> {
    const rows = await this.callproc("catalog_revision_by_number", [version]);
    if (!rows?.length) return null;
    return this.parseRevision(
      rows[0] as Record<string, unknown>,
      "PricingRepository.getBursarConfig",
    );
  }

  async activatePricing(version: number): Promise<ActivePricingRow> {
    const rows = await this.callproc("activate_catalog_revision", [version]);
    if (!rows?.length) throw new Error(`Catalog revision ${version} was not found`);
    return this.parseRevision(
      rows[0] as Record<string, unknown>,
      "PricingRepository.activatePricing",
    );
  }
}
